src/cargo/ops/cargo_test.rs

```diff
@@ -129,11 +150,22 @@ fn build_and_run<'a>(manifest_path: &Path,
         try!(config.shell().verbose(|shell| {
             shell.status("Running", cmd.to_string())
         }));
-        match ExecEngine::exec(&mut ProcessEngine, cmd) {
-            Ok(()) => {}
-            Err(e) => return Ok(Err(e))
+
+        if let Err(e) = ExecEngine::exec(&mut ProcessEngine, cmd) {
+            errors.push(e);
+            if !options.no_fail_fast {
+                break
+            }
         }
     }
-
-    Ok(Ok(compile))
+    if errors.is_empty() {
+        Ok((compile, None))
+    } else if errors.len() == 1 {
+        // Just one error occured => we can return it.
+        Ok((compile, Some(Box::new(errors.pop().unwrap()))))
+    } else {
+        // Multiple tests failed => Create a more generic error
+        let err = internal("Multiple tests failed");
+        Ok((compile, Some(err)))
+    }
```

Comment:
This looks pretty similar to the `if` statement above, perhaps the two could be refactored?